File: Java/Problems/Arrays/QuickSort.java
package Problems.Arrays;

import java.util.Arrays;

public class QuickSort {


    public void QuickSort(int[] arr, int low, int high){
        if (low<high){
            int p = partition(arr, low, high);
            QuickSort(arr, low, p-1);
            QuickSort(arr, p+1, high);
        }
    }

    private int partition(int[] arr, int low, int high) {
        int p = arr[high];
        int i=low;
        int j=high-1;

        while (true){
            while (arr[i]<p){ i++;}
            while (j>=0 && arr[j]>p){ j--;}
            if (i>=j) break;

            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
        arr[high] = arr[i];
        arr[i] = p;

        return i;
    }

    public static void main(String[] args) {
        int[] A = new int[]{5, 4, 2, 5, 3};
        new QuickSort().QuickSort(A, 0, A.length-1);
        Arrays.stream(A).forEach(i -> System.out.println("->"+i));
    }
}

File: GO-Lang/Problems/Arrays/QuickSort/QuickSort.go
package main

import "fmt"

//Given an array, sort it using quicksort.

func Sort(arr *[]int, low int, high int) {
	if low < high {
		pt := partition(arr, low, high)
		Sort(arr, low, pt-1)
		Sort(arr, pt+1, high)
	}
}

func partition(arr *[]int, low, high int) int {
	pivot := (*arr)[high]
	i, j := 0, high-1

	for {
		for (*arr)[i] < pivot {
			i++
		}
		for j >= 0 && (*arr)[j] > pivot {
			j--
		}
		if i >= j {
			break
		}
		temp := (*arr)[i]
		(*arr)[i] = (*arr)[j]
		(*arr)[j] = temp
	}
	temp := (*arr)[i]
	(*arr)[i] = pivot
	(*arr)[high] = temp
	return i
}

func main() {
	arr := []int{1, 3, 5, 7, 9, 11, 0, 4, 6, 8}
	Sort(&arr, 0, len(arr)-1)
	fmt.Println("------->", arr)
}
